### ISSUE TITLE
Terraform 0.7.10 "This is a bug with Terraform" same code works with 0.7.7

### ISSUE THREAD

**Issue by kiich** [hashicorp/terraform#10126]
Just logging this as per terraform output. Our terraform code are quite large so have not started narrowing this down.

This used to work fine with Terraform 0.7.7 and consul as remote state file.
Now using Terraform 0.7.10 and consul as remote state file (after wiping old 0.7.7 one).

### Terraform Version
```hcl
Terraform v0.7.10
```

### Expected Behavior
Terraform apply to work

### Actual Behavior
Throws this error:
```hcl
Error applying plan:

3 error(s) occurred:

* aws_route_table.routetable: diffs didn't match during apply. This is a bug with Terraform and should be reported as a GitHub Issue.

Please include the following information in your report:

    Terraform Version: 0.7.10
    Resource ID: aws_route_table.routetable
    Mismatch reason: diff RequiresNew; old: true, new: false
    Diff One (usually from plan): *terraform.InstanceDiff{mu:sync.Mutex{state:0, sema:0x0}, Attributes:map[string]*terraform.ResourceAttrDiff{"vpc_id":*terraform.ResourceAttrDiff{Old:"", New:"${var.vpc_id}", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:true, Sensitive:false, Type:0x0}, "tags.%":*terraform.ResourceAttrDiff{Old:"", New:"", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "route.#":*terraform.ResourceAttrDiff{Old:"", New:"", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}}, Destroy:false, DestroyTainted:false}
    Diff Two (usually from apply): *terraform.InstanceDiff{mu:sync.Mutex{state:0, sema:0x0}, Attributes:map[string]*terraform.ResourceAttrDiff{"tags.Name":*terraform.ResourceAttrDiff{Old:"", New:"staging-kubernetes", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "tags.owner":*terraform.ResourceAttrDiff{Old:"", New:"RUN", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "tags.Terraform":*terraform.ResourceAttrDiff{Old:"", New:"true", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "tags.vpc_id":*terraform.ResourceAttrDiff{Old:"", New:"", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "route.#":*terraform.ResourceAttrDiff{Old:"", New:"", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "tags.%":*terraform.ResourceAttrDiff{Old:"0", New:"4", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}}, Destroy:false, DestroyTainted:false}

Also include as much context as you can about your config, state, and the steps you performed to trigger this error.

* aws_subnet.subnet_in_specific_az: Error creating subnet: MissingParameter: The request must contain the parameter vpcId
	status code: 400, request id: 8d401d57-6d5c-43db-a7d4-facd563793aa
* aws_security_group.balancers: Error creating Security Group: VPCIdNotSpecified: No default VPC for this user
	status code: 400, request id: d0258c4c-4271-4a11-8a42-1412c5f5a9d5

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```

**Comment by kiich** [comment]
FYI, it breaks with 0.7.9 also:
```hcl
Error applying plan:

3 error(s) occurred:

* aws_route_table.routetable: diffs didn't match during apply. This is a bug with Terraform and should be reported as a GitHub Issue.

Please include the following information in your report:

    Terraform Version: 0.7.9
    Resource ID: aws_route_table.routetable
    Mismatch reason: diff RequiresNew; old: true, new: false
    Diff One (usually from plan): *terraform.InstanceDiff{mu:sync.Mutex{state:0, sema:0x0}, Attributes:map[string]*terraform.ResourceAttrDiff{"route.#":*terraform.ResourceAttrDiff{Old:"", New:"", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "vpc_id":*terraform.ResourceAttrDiff{Old:"", New:"${var.vpc_id}", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:true, Sensitive:false, Type:0x0}, "tags.%":*terraform.ResourceAttrDiff{Old:"", New:"", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}}, Destroy:false, DestroyTainted:false}
    Diff Two (usually from apply): *terraform.InstanceDiff{mu:sync.Mutex{state:0, sema:0x0}, Attributes:map[string]*terraform.ResourceAttrDiff{"tags.Name":*terraform.ResourceAttrDiff{Old:"", New:"staging-kubernetes", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "tags.owner":*terraform.ResourceAttrDiff{Old:"", New:"RUN", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "route.#":*terraform.ResourceAttrDiff{Old:"", New:"", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "tags.%":*terraform.ResourceAttrDiff{Old:"0", New:"4", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "tags.Terraform":*terraform.ResourceAttrDiff{Old:"", New:"true", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "tags.vpc_id":*terraform.ResourceAttrDiff{Old:"", New:"", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}}, Destroy:false, DestroyTainted:false}

Also include as much context as you can about your config, state, and the steps you performed to trigger this error.

* aws_subnet.subnet_in_specific_az: Error creating subnet: MissingParameter: The request must contain the parameter vpcId
	status code: 400, request id: 33adacbd-52e9-4181-be77-12a541def014
* aws_security_group.balancers: Error creating Security Group: VPCIdNotSpecified: No default VPC for this user
	status code: 400, request id: 692aa6e4-1aff-4ba0-8f4d-68db6150e98a

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```

**Comment by mitchellh** [comment]
Dup of #9916. We've been trying to track this down but need a configuration+state to reproduce this. Is there any chance you can get this?

**Comment by kiich** [comment]
@mitchellh sure - here's the link https://gist.github.com/kiich/3edc859f91241528db7e9dc5ed1e87f5
for all files involved.

It's a vastly cut down version of what we use but it does reproduce the error.
few things to note - we started of using TF modules which simply outputs a vpc id from module's variable file before we started looking at remote state file hence we use the template file in a bit of weird logic.

basically, when the module gets passed certain variable, the module works out whether it needs to:

- create a new VPC and output its ID
- output an existing ID

and we use the template heavily to achieve that.

i've replaced sensitive info, hopefully you can still reproduce the problem!

**Comment by kiich** [comment]
I understand also that I am using deprecated feature template_file - i haven't tried it with data source yet.

**Comment by kiich** [comment]
tried with data source template_file - i can still reproduce the issue.

**Comment by mitchellh** [comment]
Reopening, I haven't tested yet but will get to that.

**Comment by jbardin** [comment]
I think I see the problem here. The interpolation in the template needs to be escape. THere's better detection for this added recently (forget which PR), but you should now see the error:

```
  * invalid variable syntax: "vpc_id_from_map". If this is part of inline `template` parameter
then you must escape the interpolation with two dollar signs. For
example: ${a} becomes $${a}.
```

Changing the template to `template = "$${vpc_id_from_map}"` prevents the diff mismatch during apply

**Comment by mitchellh** [comment]
That's right! There is detection in 0.8, but we didn't backport to 0.7 since that represents a BC-incompat change

**Comment by kiich** [comment]
Nice, that's good to know - I'll test the fix in our repo first thing but
looks like this is the cause.

Thanks for the quick response/resolution guys!

**Comment by kiich** [comment]
Just to let everybody know - the fix was indeed:
```template = "$${vpc_id_from_map}"```
i.e. escaping using two dollar sign.

**Comment by ndp-stc** [comment]
A line number would have been nice.
